libminifi/include/utils/ChecksumCalculator.h

```diff
@@ -0,0 +1,57 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <string>
+#include <utility>
+
+#include "utils/OptionalUtils.h"
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+class ChecksumCalculator {
+ public:
+  static constexpr const char* CHECKSUM_TYPE = "SHA256";
+  static constexpr size_t LENGTH_OF_HASH_IN_BYTES = 32;
+
+  void setFileLocation(const std::string& file_location);
+  std::string getFileName() const;
+  std::string getChecksum();
+  void invalidateChecksum();
```

Comment:
Did you consider `stat`-ing the files on each call and recalculating the checksum based on the last modified timestamp? This would fix the case when the watched files were modified outside of minifi. An alternative faster, but way harder and more complex approach would be subscribing to filesystem events like `inotify` (Linux)/`FindFirstChangeNotification`(Windows)/`fsevents`(Mac OS X), and recalculating the hashes or invalidating the cache on change events.

Reply:
I decided not to do that because minifi does not automatically re-read the configuration if it is changed on the disk.  We want to send the checksum of the configuration files as currently used by minifi; this can only change at startup and when minifi itself modifies it in response to a C2 command.

For example, if minifi.properties is modified but the change is reverted before the next minifi restart, sending a modified checksum would be a false alert.

Reply:
That's fair, thanks